jwt/jwt.c

```diff
@@ -83,21 +85,34 @@ char* jwt_create(char * project_id, long long int time, NN_DIGIT * priv_key, int
 
 int jwt_verify(char * jwt,  point_t * pub_key)
 {
-    ecc_init(); // TODO: Has to be initialized only once -> move to broker init phase
-
-    char * chunk_head = jwt;
-    char * dot_pos = strchr(chunk_head, '.');
-    simple_array_t jwt_header = base64_decode(chunk_head, dot_pos - chunk_head + 1);
-    chunk_head = dot_pos + 1;
-    dot_pos = strchr(chunk_head, '.');
-    simple_array_t jwt_body = base64_decode(chunk_head, dot_pos - chunk_head + 1);
-    chunk_head = dot_pos + 1;
-    simple_array_t jwt_signature = base64_decode(chunk_head, strlen(chunk_head));
+    //ecc_init(); // TODO: Has to be initialized only once -> move to broker init phase
+
+    // First check if jwt is in correct format then decode
+    char * dot_pos_1 = strchr(jwt, '.');
+    if (!dot_pos_1) {
+        mosquitto_log_printf(MOSQ_LOG_ERR, "Invalid JWT: Missing dot!");
+        return 0;
+    }  // dot_pos_1 = header_end
+    
+    char * dot_pos_2 = strchr(dot_pos_1 + 1, '.');
+    if (!dot_pos_2) {
+        mosquitto_log_printf(MOSQ_LOG_ERR, "Invalid JWT: Missing dot!");
+        return 0;
+    }  // dot_pos_2 = body_end
+
+    if (strlen(dot_pos_2 + 1) < (2 * NUMBYTES)) {
+        mosquitto_log_printf(MOSQ_LOG_ERR, "Invalid JWT: Signature too short!");
+        return 0;
+    }  // check signature length
+
+    simple_array_t jwt_header = base64_decode(jwt, dot_pos_1 - jwt + 1);
```

Comment:
Please double check these calculations.  The first one seems incorrect (extra +1)
Example: 
1f.
0 1 2
jwt = 0
dot_post = 2

lenght = 2 - 0 = 2

Reply:
You are right, I was adding extra +1 to header and body decoding functions. I fixed it.